### ISSUE TITLE
Single Worker thread, single I/O thread

### ISSUE THREAD

**Issue by eliezedeck** [dotnet/corefx#13586]
Hey guys,

I have been working on a project that involves async/await a lot. And with the multiple worker threads and I/O thread, the order at which operations happen are pretty much mangled together, and things just go wrong all the time. Because of this too, it's really hard to debug what's happening.

I'm looking for a way to **limit both the Worker and I/O threads** to both be just 1. I need this, because it's easier this way to maintain the application, easier to reason about and debug, and save a number of threads along the way.

Admittedly, I reasoned wrong with the .Net Core's async/await, which is mostly always single-threaded in other languages; so, I assumed it would also be single-threaded on .Net, but I guess I assumed completely wrong, my bad. But it's been a bad surprise for me to realize that worker threads are used by C# to make async/await happen. Now, unless things are not causing races, the project is useless now.

Please suggest what do I do to make it single-threaded.
Thanks

**Comment by svick** [comment]
A way to do this might be to run your application on a single-threaded `SynchronizationContext`. That way, any `await` that does not use `.ConfugureAwait(false)` will resume on that thread.

A simple way to do that is to use [`AsyncContext` from AsyncEx](https://github.com/StephenCleary/AsyncEx/wiki/AsyncContext).

**Comment by SunnyWar** [comment]
@eliezedeck I'm confused. Async/Await is not multi-threaded. It does not create additional worker threads. [https://msdn.microsoft.com/en-us/library/hh191443(v=vs.110).aspx](url)  

> The async and await keywords don't cause additional threads to be created. Async methods don't require multithreading because an async method doesn't run on its own thread. The method runs on the current synchronization context and uses time on the thread only when the method is active. You can use Task.Run to move CPU-bound work to a background thread, but a background thread doesn't help with a process that's just waiting for results to become available.

**Comment by alexperovich** [comment]
@SunnyWar async/await doesn't have to be multithreaded, but if the SynchronizationContext that the Task is running on provides multiple threads (like the thread pool, which is used by default) then the async/await execution will be multithreaded.

@svick's solution is the correct one. If you need a sequence of async/await to be single threaded you must use a single threaded SynchronizationContext.

**Comment by eliezedeck** [comment]
@svick Thanks. All async operations don't have to be awaited, like the `async void` which returns no Task but will do async work eventually. At times, you explicitly want this, and that is a requirement.

Consider this situation: I do a lot of network async operations, I need a deadline to a group of operations. I obviously need to spin at least one Task.Delay() before all operations begin, which will manage the deadline and do some logic later on.

**Comment by Clockwork-Muse** [comment]
@eliezedeck - wouldn't it be cleaner to do something like:
```
CancellationToken deadline = new CancellationTokenSource(timeout).Token;
Task[] tasks = actions.Select(x => Task.Run(x, deadline)).ToArray();
Task completed = Task.WhenAll(tasks);
await completed;
// Or maybe doing something with completed.ContinueWith(...)
```
(there are potential problems with the above code, but you get what I mean)

My gut feeling is that if you're dealing with async code and expecting only one thread, or adding explicit delays to try and match things up, you may be able to change how you structured your code to take better advantage of things.

**Comment by eliezedeck** [comment]
I guess I gave a pretty naive example. And I have to admit, this is really the very first .Net project that I'm working on, and I find the language pretty cool.

@Clockwork-Muse May be I'm not yet a very good C# developer. But my usecase involves a lot of values being modified at each end of the await(s), and since they are not known at the beginning of the program, I don't think I can make use of that, and it's a bit weird for me to reason like that, since it's not like a simple increment operation, it's a long chain of operations, many counters and values updated along the way.

@svick The Nito.AsyncEx doesn't work on .Net Core, I think it is only compatible with .Net 4.5 (and earlier). Instead, I'm using `Shaman.SingleThreadSynchronizationContext` and things seem to be behaving as expected. There are pretty low download counts on that package, anyone care to comment on its correctness and quality?

One thing though, I'm still seeing other threads in the debugger when I pause the app. This leads to another question, are Task.Delay() based on sleeping threads, or I/O works that the kernel will notify us eventually? Because the project will definitely make use of lots of timers, because there are lots of clients involved. If there is going to be a Thread for each `await Task.Delay(...)`, even though they are thread pool, I think that's not going to scale. May be I'm wrong, but I'm used to thinking about the `timer_fd` on Linux, and I know I can have many of that in one app without impacting the scalability.

**Comment by svick** [comment]
@eliezedeck 

There are .Net Core versions of AsyncEx types. For `AsyncContext`, it's in [AsyncEx.Context](https://github.com/StephenCleary/AsyncEx.Context).

And `Task.Delay()` certainly does not use a sleeping thread, instead it uses a timer. (You can verify that by looking at [the source code](https://github.com/dotnet/coreclr/blob/19346e5/src/mscorlib/src/System/Threading/Tasks/Task.cs#L5881).)

**Comment by alexperovich** [comment]
@eliezedeck There will always be at least two threads for a .net application, the main thread and the GC finalizer thread. For most applications there will also be one or more threads for the ThreadPool.

I am curious what your full use case is. Normally a single threaded context is only needed if you have a UI thread or STA COM objects. If you await things properly you will get a single chain of execution. The callbacks will not necessarily run on the same thread but they will execute sequentially.

**Comment by eliezedeck** [comment]
@alexperovich It's a proxy that uses all other non-TCP protocols as the main transport.

- Accept connection
- start reading TCP using async (this is not awaited)
- split data in order to send to Transports
- each transport has their own speed, depending on the Protocol available, and `await Task.Delay()` is used to delay sending in order to avoid overwhelming each transport

I agree I may have the design fundamentally wrong, or I should use `await` in some more places. I'm happy to received corrections to my design. Like I said, I come from a different platform where things are done a bit differently, and that may affect all my design.

**Comment by jnm2** [comment]
@eliezedeck I realize this is not the same as what you are doing but I wonder if you would find it interesting to compare. It's a simple full-duplex TCP proxy that has been solid in production. https://github.com/jnm2/CoreProxy/blob/master/src/CoreProxy/Proxies/TcpProxyService.cs#L115

Also, I agree that your code probably shouldn't have to know or care that it's running on multiple threads. If you have multiple concurrent tasks, you use `await Task.WhenAll(...)` to continue on a single thread when they are all finished. It shouldn't matter which thread.

**Comment by eliezedeck** [comment]
@jnm2 Thanks for your input, I'll look into that code for insights.

The problem was not about which Thread did which task, it was about **multiple threads making changes to the same instance** of a class at the same time and causing races. That was the reason behind getting a single-threaded `SynchronizationContext`.

I know this could have also been solved by using locks, but really, it makes the code ugly, and harder to read with all that `lock () {...}`. And there are several members being updated every-time, and there are also at least 4 classes involved having to be updated also. Now, you can imagine the amount of `lock` I'll have to write everywhere.

**Comment by jnm2** [comment]
@eliezedeck I'm curious to see your code to get a clearer picture but I'm guessing you have multiple transports finishing concurrently and you need to merge their results or stats into one place without one waiting for the others to finish.

 * If that's the case, you're right. I think you'll need a synchronization context or a synchronization primitive such as `lock`. (I have a hard time visualizing why you'd need more than one lock statement total across all the code. I'm thinking the class instance can be thread-safe by making all properties readonly and mutating it only by sending parameters to a single method. That method would contain the lock statement.)
 * If however you can wait until all in-flight operations have stopped before centrally updating the results and stats, you can use `await Task.WhenAll` instead of locking. When you said "deadline" this is what I was picturing.

**Comment by eliezedeck** [comment]
@jnm2 It's the first case indeed. There are several members, like `sentCount`, `sentBytes`, `unackedCount`, `resentCount`, etc... these get updated in completely different contexts.

Now, there are a lot of inter-dependencies, and some operations that start depends on one value, but suddenly the value is changed by some other threads: nightmare. Apparently, the Single threaded synchronization solves the problem, but I'll have to dig more. At this time, there is a situation that the single thread is locked in a loop forever, and nothing happens.

I'm guessing async/await in .Net really do need multiple threads to work right; anyway, it was design to be so, and I'm fighting with it. I guess I'll have to use multiple threads in the end, and will have to change my codes to use locks.

**Comment by jnm2** [comment]
This might be the cause: `.Result` and `.Wait()` and `.GetAwaiter().GetResult()` cause deadlocks on a task that hasn't ended yet when you call from the same thread which the task's continuation must be synchronized on. On a single-thread sync context, you'll have to await everywhere and only call the above once you know the task is complete.

If it was me I'd go multi-threaded which is less overhead, put all the stats you mentioned as readonly properties in a single class, and add class method `AddStats(sentCount, sentBytes, unackedCount, resentCount)`. The entire method body should be in a `lock` statement. Can you see any issues with taking that approach?

**Comment by eliezedeck** [comment]
@jnm2 I don't use those `.Wait()` (...) methods that you mentioned. And as far as I understand, there should be no other threads able to run while the current one is running. Wasn't that the essence of the single-threaded sync context? It makes absolutely no sense to me to use it while there is still a rogue thread calling stuffs on my code without me knowing.

But I'll definitely change the approach. Seeing that C#'s async/await are designed around multi-threaded at every level, I think that is how I should design my app too.

Thanks everyone for all your answers. I'll close this ticket if there is no suggestions anymore. It was a good learning.

**Comment by jnm2** [comment]
@eliezedeck On a single-threaded context, it *does* make no sense; thus the deadlock. If you don't use `Task.Result` or `.Wait()` then ignore that. 

C#'s async and await is used extensively in an single-threaded manner. When you write a Windows Forms app or a WPF app, everything runs on the UI thread. It's just not normal for a non-UI app, but it *should* be mature and work.

**Comment by alexperovich** [comment]
Closing as the question has been answered. If you have another question please make a new issue.
